### PR TITLE
refactor(geometry): Surface merging returns ordering

### PR DESCRIPTION
**Pull request by paulgessinger**
This PR updates the surface merging for Cylinders and Discs:

1. Surface merging returns their relative "ordering", e.g. for two cylinders `a` and `b`, the merging will return whether a was at lower $z$ or $r\phi$ values than `b`. The same applies for discs. For merged surfaces in $\phi$ or $r\phi$ direction which cover the full azimuth after merging, the ordering is preserving: `a` + `b` will return `a`, `b`, and `b` + `a` will return `b`, `a` accordingly. 	
2. Cylinders and discs support *external rotation*: with this option set, the merged surface will always have an average $\phi = 0$, meaning that no additional rotation is applied on top of the rotation by the transform. This is needed for well-defined portal merging
3. Remove the geometry context argument from merging. Surface merging is strictly only supported for surfaces without detector elements, which is checked by the methods. After this is established, it's safe to access the nominal transform directly, without using the geometry context mechanism (it won't be used anyway without detector elements).
4. Improved coverage of test cases, including a number of edge cases, and associated bug fixes.

### CODE REVIEW

File: Core/include/Acts/Surfaces/Surface.hpp [hunk 1]
```diff
@@ -528,4 +528,6 @@ class Surface : public virtual GeometryObject,
       const Vector3& direction) const;
 };
 
+std::ostream& operator<<(std::ostream& os, const Surface& srf);
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/include/Acts/Surfaces/Surface.hpp:531]: Oh yeah. I'll remove this.

File: Tests/UnitTests/Core/Surfaces/DiscSurfaceTests.cpp [hunk 2]
```diff
@@ -576,50 +618,116 @@ BOOST_DATA_TEST_CASE(PhiDirection,
...
+  }
 
-  BOOST_CHECK_SMALL(
-      detail::difference_periodic(bounds45->get(RadialBounds::eAveragePhi),
-                                  a(180_degree), 2 * M_PI),
-      1e-6);
-  BOOST_CHECK_CLOSE(bounds45->get(RadialBounds::eHalfPhiSector), 30_degree,
-                    0.1);
+  // @TODO: Add half-circle sector test
+  // @TODO: Add detector element test
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Tests/UnitTests/Core/Surfaces/DiscSurfaceTests.cpp]: They were indeed still missing, I added these tests now.

**Review by andiwand — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :red_circle: Athena integration test results [cc3f620234336a97c4fe5ea0c252fad24e6bee64]
### :red_circle: Some tests have failed!
Please investigate the [pipeline](https://gitlab.cern.ch/acts/acts-athena-ci/-/pipelines/7898558)!

|status|job|report|
|------|---|------|
|:green_circle: | [report_pull_request](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126265)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126264)||
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126263)||
|:green_circle: | [test_ActsDumpGeometryIdentifiers](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126262)||
|:red_circle: | [test_ActsCheckObjectCountsCachedWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126261)||
|:red_circle: | [test_ActsCheckObjectCountsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126260)||
|:green_circle: | [test_ActsEFTrackFit](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126259)||
|:green_circle: | [test_ActsPersistifySeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126258)||
|:green_circle: | [test_ActsBenchmarkWithSpot](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126257)||
|:green_circle: | [test_ActsAnalogueClustering](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126256)||
|:green_circle: | [test_ActsWorkflowHeavyIons](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126255)||
|:green_circle: | [test_ActsWorkflowFastTracking](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126254)||
|:green_circle: | [test_ActsWorkflowCached](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126253)||
|:green_circle: | [test_ActsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126252)||
|:green_circle: | [test_ActsValidateAmbiguityResolution](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126251)||
|:green_circle: | [test_ActsValidateResolvedTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126250)||
|:green_circle: | [test_ActsValidateTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126249)||
|:green_circle: | [test_ActsValidateActsCoreSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126248)||
|:green_circle: | [test_ActsValidateActsSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126247)||
|:green_circle: | [test_ActsValidateSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126246)||
|:green_circle: | [test_ActsValidateOrthogonalSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126245)||
|:green_circle: | [test_ActsValidateClusters](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126244)||
|:green_circle: | [test_ActsPersistifyEDM](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126243)||
|:green_circle: | [test_ActsGSFRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126242)||
|:green_circle: | [test_ActsKfRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126241)||
|:green_circle: | [test_ActsExtrapolationAlgTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126240)||
|:green_circle: | [test_ActsITkTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126239)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126238)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126237)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126236)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126235)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126234)||
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42126233)||
